LFS-119: As a user, select date-time and date-time ranges as answers to questions

OrdiNeu opened the pull request:
[LFS-119](https://phenotips.atlassian.net/browse/LFS-119).

Test rig this time: (See 8bfbf09058f01292a53e553b8f480a8491a330b6 )

`modules/data-entry/src/main/frontend/webpack.config.js` line 19:
```
    [module_name + 'testRig']: './src/questionnaire/testRig.jsx'
```

`modules/data-entry/src/main/frontend/src/questionnaire/testRig.jsx` line 19:
```
//
//  Licensed to the Apache Software Foundation (ASF) under one
//  or more contributor license agreements.  See the NOTICE file
//  distributed with this work for additional information
//  regarding copyright ownership.  The ASF licenses this file
//  to you under the Apache License, Version 2.0 (the
//  "License"); you may not use this file except in compliance
//  with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing,
//  software distributed under the License is distributed on an
//  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
//  KIND, either express or implied.  See the License for the
//  specific language governing permissions and limitations
//  under the License.
//

import React from "react";

import DateQuestion from "./DateQuestion";
import NumberQuestion from "./NumberQuestion";

import moment from "moment";

export default function TestRig(props) {
  return (
    <React.Fragment>
      <DateQuestion
        name="Please enter your date of birth"
        precision="yyyy-MM-dd"
        type="timestamp"
        />
      <DateQuestion
        name="Year-only input"
        precision="yyyy"
        type="timestamp"
        />
      <DateQuestion
        name="Year+month input"
        precision="yyyy-MM"
        type="timestamp"
        />
      <DateQuestion
        name="Please enter a date range in 2019"
        precision="yyyy-MM-dd"
        lowerLimit={moment("2019-01-01")}
        upperLimit={moment("2019-12-31")}
        type="interval"
        />
      <DateQuestion
        name="Please enter a date-time in 2019"
        precision="yyyy-MM-dd HH:mm:ss"
        lowerLimit={moment("2019-01-01")}
        upperLimit={moment("2019-12-31")}
        type="timestamp"
        />
      <DateQuestion
        name="Please enter a year range"
        precision="yyyy"
        type="interval"
        />
    </React.Fragment>);
}
```

`modules/data-entry/src/main/resources/SLING-INF/content/Extensions.json` line 41:
```
"TestRig2": {
        "jcr:primaryType": "lfs:Extension",
        "lfs:extensionPointId": "lfs/coreUI/sidebar/entry",
        "lfs:targetURL": "TestRig",
        "lfs:extensionName": "Test rig",
        "lfs:icon": "asset:lfs-dataentry.questionnairesIcon.js",
        "lfs:extensionRenderURL": "asset:lfs-dataentry.testRig.js",
        "jcr:content": {
            "jcr:primaryType": "nt:resource",
            "jcr:data": "",
            "jcr:mimeType": "text/html"
        }
    }
```

OrdiNeu committed:
LFS-119: As a user, select date-time and date-time ranges as answers to questions
Add a DateQuestion widget to handle questions where the answers are
dates, date-times, or intervals

OrdiNeu committed:
LFS-119: As a user, select date-time and date-time ranges as answers to questions
Enable the "interval" type, which allows for two inputs where the first
input must be before the second
Move state from MultipleChoice to NumberQuestion when a range is
provided, in order to make error checking work across multiple inputs.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -40,6 +40,14 @@ const questionnaireStyle = theme => ({
       paddingBottom: "0px",
       paddingTop: theme.spacing(1),
     },
+    answerPadding: {
```
`answerPadding` is a visual description. Is there a more semantic term that could be used to designate components that need this padding?

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestion.jsx:
```diff
@@ -0,0 +1,236 @@
+          className={classes.textField}
+          InputLabelProps={{
+            shrink: true,
+            max: upperLimit,
```
`max` and `min` should be `InputProps`, not `InputLabelProps`.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestion.jsx:
```diff
@@ -0,0 +1,236 @@
+          max: upperLimit,
+          min: lowerLimit
+        }}
+        onChange={
```
This doesn't seem to really work in Firefox, the selected value is not bound to the min-max interval.

OrdiNeu replied:
Just fixed it -- there was an issue where the Date() object was being parsed properly in Chrome but not Firefox. CHeck the update to testRig.jsx

sdumitriu reviewed (changes requested):
In Firefox this is really clunky. I guess it's OK for the moment, but we should really try to use a more powerful datetime picker widget.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestion.jsx:
```diff
@@ -0,0 +1,236 @@
+
+const DATETIME_FORMATS = [
+  "yyyy-MM-dd hh",
+  "yyyy-MM-dd hh:mm",
```
Oh, the first part of the comment is kind of wrong, moment also uses `h` and `H`, so it has to be `HH` anyway. I've made the same mistake of using `h` instead of `H` in the past as well.

sdumitriu commented:
> In Firefox this is really clunky. I guess it's OK for the moment, but we should really try to use a more powerful datetime picker widget.

The worst thing is that for the datetime field, there's no picker, and trying to type anything in it other than carefully replacing one digit at a time leads to `Invalid date` as the input value, which can't be changed other than pasting a valid ISO date.

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestion.jsx:
```diff
@@ -0,0 +1,236 @@
+          }}
+          onChange={
+            (value) => {
+              changeEndDate(boundEndDate(amendMoment(event.target.value, precision), selectedDate));
```
`boundEndDate()` uses `selectedDate` (i.e. the start of the interval) as its second argument in order to make sure that the end is always after the beginning. Here, the `selectedEndDate` I'm setting is in `event.target.value` -- I'll make this clearer by refactoring it.

sdumitriu replied:
Yes, I figured this out eventually but I forgot to delete the pending comment.

OrdiNeu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestion.jsx:
```diff
@@ -0,0 +1,236 @@
+]
+
+const DATETIME_FORMATS = [
+  "yyyy-MM-dd hh",
```
I use them in PropType checks, using the `oneOf()` syntax

sdumitriu replied:
I didn't mean all of them just the one with date+hour.

OrdiNeu committed:
LFS-119: As a user, select date-time and date-time ranges as answers to questions
Remove a useless Datetime format and fix the rest.
Speed up/clean date truncation.
Make precision optional with a default.
Fix invalid default prop.
Code cleanup.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -39,6 +39,15 @@ const questionnaireStyle = theme => ({
       display: 'inline-block',
       paddingBottom: "0px",
       paddingTop: theme.spacing(1),
+      with: "100%"
```
```suggestion
      width: "100%"
```

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/QuestionnaireStyle.jsx:
```diff
@@ -39,6 +39,15 @@ const questionnaireStyle = theme => ({
       paddingTop: theme.spacing(1),
+      with: "100%"
+    },
+    answerPadding: {
```
This isn't used anymore.

OrdiNeu replied:
I'm still in the middle of edits -- the width:100% is being problematic because I have to apply it to the search wrapper + the InputProps + the inputProps. Will push the rest of my changes once I get this fixed.

OrdiNeu committed:
LFS-119: As a user, select date-time and date-time ranges as answers to questions
Fix styling issues with the width of inputs

OrdiNeu committed:
LFS-119: As a user, select date-time and date-time ranges as answers to questions
Update answers to use the end date (if necessary)

OrdiNeu committed:
LFS-119: As a user, select date-time and date-time ranges as answers to questions
Fix date limits & key warning.

OrdiNeu committed:
LFS-119: As a user, select date-time and date-time ranges as answers to questions
Forgot a package dependency.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaire/DateQuestion.jsx:
```diff
@@ -0,0 +1,252 @@
+// Optional props:
+// name: the question to be displayed
+// type: "timestamp" for a single date or "interval" for two dates
+// precision: yyyy, yyyy-MM, yyyy-MM-dd, yyyy-MM-dd hh,  yyyy-MM-dd hh:mm, yyyy-MM-dd hh:mm:ss
```
```suggestion
// precision: yyyy, yyyy-MM, yyyy-MM-dd, yyyy-MM-dd hh:mm, yyyy-MM-dd hh:mm:ss
```

OrdiNeu committed:
LFS-119: As a user, select date-time and date-time ranges as answers to questions

Documentation fix in modules/data-entry/src/main/frontend/src/questionnaire/DateQuestion.jsx by @sdumitriu

Co-Authored-By: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

OrdiNeu committed:
LFS-119: As a user, select date-time and date-time ranges as answers to questions
Fix themeing and improper truncation

marta- reviewed (approved):
Looks good. The UI for date intervals needs some work, but not in this PR. A separate task will be created for it.

sdumitriu commented:
One thing to keep in mind: when passing `lowerLimit` and `upperLimit`, `new Date(...)` should not be used, because it parses them as UTC timestamps that are then shifted to the local timezone, which means that one could select `2018-12-31` for a 2019-only date, but can't select `2019-12-31`.
